test/controllers/products_controller_test.rb

```diff
@@ -0,0 +1,135 @@
+require "test_helper"
+require "pry"
+
+describe ProductsController do
+  describe "index" do
+    it "succeeds when there are works" do
+        Product.count.must_equal 3
+        get products_path
+        must_respond_with :success
+    end
+
+    it "succeeds when there are no works" do
+        Product.all.each do |item|
+          item.destroy
+        end
+
+        Product.count.must_equal 0
+        get products_path
+        must_respond_with :success
+      end
+  end
+
+  describe "show" do
+    it "succeeds" do
+      get product_path(products(:cat))
+      must_respond_with :success
+    end
+
+    # it "redirects if not in db" do
+    #   get product_path(nil)
+    #   must_redirect_to products_path
+    # end
+  end
+
+  describe "new" do
+    it "succeeds" do
+      user = users(:ada)
+      get new_user_product_path(user.id)
+      must_respond_with :success
+    end
+  end
+
+  describe "create" do
+    it "creates a product with valid data" do
+      Product.count.must_equal 3
+            user = users(:ada)
+
+            post user_products_path(user.id), params: {
+              product: {
+                name: "Bunny",
+                stock: 14,
+                price: 5,
+                description: "Too many, please help",
+                status: true,
+                user_id: user.id,
+                image_url: "www.test-URL.com",
+                categories: [categories(:one)]
+              }
+            }
+      Product.count.must_equal 4
+      must_respond_with :redirect
+      must_redirect_to products_path
+    end
+
+    it "won't create invalid product" do
```

Comment:
This test is failing because the view is trying to access product fields for a nil product.  :(